### PR TITLE
Add ticks + zoom scaling

### PR DESCRIPTION
**Pull request by andyrichardson**
## About

- Add ticks to timeline
- Add zooming and panning (scaled)
- Trim height of timeline-row durations

**[COSMOS](https://deploy-preview-114--urql-devtools.netlify.com/_renderer.html?_fixtureId=%7B%22path%22%3A%22src%2Fpanel%2Fpages%2Ftimeline%2FTimeline.fixture.tsx%22%2C%22name%22%3A%22basic%22%7D)**

## Notes
 - Does scroll + zoom feel intuitive/obvious?

### CODE REVIEW

File: src/panel/context/Explorer/ast/index.ts [hunk 1]
```diff
@@ -1,177 +1,221 @@
...
-  request: Operation,
-  data: Data,
-  map: NodeMap = Object.create(null)
-) => {
-  if (request.operationName !== "query") {
-    return map;
+export const handleResponse = ({
+  operation,
+  data,
+  parsedNodes = {}
```
Comment (wgolledge) [inline src/panel/context/Explorer/ast/index.ts]: I'm assuming we're not so concerned with bundlesize here in `devtools`? Also can see the logic in sticking with undefined consistently if using default params 👍

File: src/panel/context/Explorer/ast/index.ts [hunk 2]
```diff
@@ -1,177 +1,221 @@
...
-    variables: normalizeVariables(operation, request.variables),
-    fragments: getFragments(request.query),
-    cacheOutcome: request.context.meta && request.context.meta.cacheOutcome
-  };
+  const fragments = operation.query.definitions
+    .filter(isFragmentNode)
+    .reduce<Record<string, FragmentDefinitionNode>>(
+      (map, node) => ({
+        ...map,
+        [node.name.value]: node
```
Comment (wgolledge) [inline src/panel/context/Explorer/ast/index.ts]: Will this always exist?

File: src/panel/context/Explorer/ast/index.ts [hunk 3]
```diff
@@ -1,177 +1,221 @@
...
-        delete node.children;
-      }
-    } else {
-      const fragmentNode = !isInlineFragment(fieldNode)
-        ? ctx.fragments[getName(fieldNode)]
-        : fieldNode;
-      if (fragmentNode !== undefined) {
-        copyFromData(ctx, map, getSelectionSet(fragmentNode), data, owner);
-      }
+    if (
```
Comment (wgolledge) [inline src/panel/context/Explorer/ast/index.ts]: Could it be helpful to pull this condition out into a var/func to help describe what's happening?

File: src/panel/context/Explorer/ast/index.ts [hunk 4]
```diff
@@ -1,177 +1,221 @@
...
+      selectionNode.selectionSet &&
+      !Array.isArray(value) &&
+      value &&
+      typeof value === "object"
+    ) {
+      return {
+        ...parsedNodemap,
+        [key]: {
+          ...node,
+          children: parseNodes({
```
Comment (wgolledge) [inline src/panel/context/Explorer/ast/index.ts]: I guess the children prop could be done with a ternary here as with `parseNodes` above

File: src/panel/context/Explorer/ast/variables.ts [hunk 5]
```diff
@@ -1,87 +1,41 @@
...
-  }, Object.create(null));
-};
+export const getNormalizedVariables = (
+  variableDefinitions: OperationDefinitionNode["variableDefinitions"] = [],
+  variables?: Maybe<Record<string, any>>
+) =>
+  variableDefinitions.reduce<Operation["variables"]>(
+    (normalized, definition) => ({
+      ...normalized,
+      [definition.variable.name.value]: valueFromASTUntyped(
```
Comment (wgolledge) [inline src/panel/context/Explorer/ast/variables.ts]: This is totally fine as is, but I guess we could also use a `Map` in instances like this?

File: src/panel/context/Timeline.tsx [hunk 6]
```diff
@@ -41,10 +47,29 @@ const useTimelineDomain = () => {
...
+    const endTime = domain.current.zoom * 30000;
+
+    const newScale = scaleLinear()
+      .domain([domain.current.start, endTime])
+      .range([0, ref.current.clientWidth]);
+
+    setScale(oldScale => {
+      if (
+        oldScale.scale &&
+        // Scale hasn't changed
```
Comment (wgolledge) [inline src/panel/context/Timeline.tsx]: 👍

File: src/panel/pages/explorer/components/Icons.tsx [hunk 7]
```diff
@@ -35,20 +29,10 @@ export const SeeMoreIcon = ({ className }: { className?: string }) => (
...
-  state: "hit" | "miss" | "partial" | undefined;
-}) => (
-  <svg
-    width="10"
-    height="10"
-    viewBox="0 0 10 10"
-    fill="none"
-    className={className}
-  >
+export const CacheOutcomeIcon: FC<JSX.IntrinsicElements["svg"] & {
```
Comment (wgolledge) [inline src/panel/pages/explorer/components/Icons.tsx]: Type wizard

File: src/panel/pages/timeline/Timeline.tsx [hunk 8]
```diff
@@ -1,39 +1,58 @@
...
-      </Container>
+      </Page>
     );
 
   return (
-    <Container>
+    <Page>
       <TimelineList ref={setContainer}>
+        {ticks.map(t => (
+          <Tick
```
Comment (wgolledge) [inline src/panel/pages/timeline/Timeline.tsx]: From the useMemo above it looks like ticks could be undefined - do we need to a truthy check here or will it be okay?

File: src/panel/pages/timeline/Timeline.tsx [hunk 9]
```diff
@@ -1,39 +1,58 @@
...
 import { useTimelineContext } from "../../context";
 import { Background } from "../../components/Background";
-import { TimelineRow, TimelinePane } from "./components";
+import { TimelineRow, TimelinePane, Tick } from "./components";
 
 export const Timeline: FC = () => {
   const {
     setContainer,
-    getTimePosition,
+    scale,
```
Comment (wgolledge) [inline src/panel/pages/timeline/Timeline.tsx]: Getting some type errors here

File: src/panel/pages/explorer/components/Icons.tsx (hunk 7 above)
Comment (andyrichardson) [reply src/panel/pages/explorer/components/Icons.tsx]: I did this because we could then animate the icon rotating

Comment (andyrichardson) [reply src/panel/pages/explorer/components/Icons.tsx]: Didn't get around to animating yet!

**Comment by wgolledge** [comment]
Looking awesome in cosmos!

One slightly strange interaction - I get a horizontal scroll if I vertical scroll when already at the bottom of the container/domain

File: .eslintrc [hunk 10]
```diff
@@ -10,8 +10,15 @@
     "plugin:react/recommended"
   ],
   "rules": {
-    "import/order": 2,
-    "@typescript-eslint/no-unused-vars": 2,
+    "import/order": ["error", { "newlines-between": "never" }],
```
[the remark this answers is not in the document]
Comment (andyrichardson) [reply .eslintrc]: Are you saying we need to order the import order rule? 😅

File: src/panel/context/Explorer/ast/index.ts (hunk 1 above)
Comment (andyrichardson) [reply src/panel/context/Explorer/ast/index.ts]: Yeah I'm not too concerned about bundle size as assets are locally stored when installing the extension

File: src/panel/pages/timeline/Timeline.tsx (hunk 8 above)
Comment (andyrichardson) [reply src/panel/pages/timeline/Timeline.tsx]: Good shout - fixed!

File: src/panel/pages/timeline/Timeline.tsx (hunk 9 above)
Comment (andyrichardson) [reply src/panel/pages/timeline/Timeline.tsx]: I think the TS server is using old values - it's passing type checking now 🙏

**Comment by andyrichardson** [comment]
> One slightly strange interaction - I get a horizontal scroll if I vertical scroll when already at the bottom of the container/domain

@wgolledge we should pair on this!

The intention was to allow horizontal scrolling when using a mouse - but this wasn't my brightest of ideas. Lets try out dragging like you suggested!

**Pull request merged**